nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -0,0 +1,441 @@
+            echo "Configuration application failed, retrying.. ($attempt/5)" >&2
+            sleep 3
+        done
+
+        cp configuration/standalone.xml $out
+      '';
+    in
+      lib.mkIf cfg.enable {
+
+        environment.systemPackages = [ cfg.package ];
+
+        systemd.services.keycloakDatabaseInit = lib.mkIf databaseActuallyCreateLocally {
+          after = [ "postgresql.service" ];
+          before = [ "keycloak.service" ];
+          bindsTo = [ "postgresql.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            RemainAfterExit = true;
+            User = "postgres";
+            Group = "postgres";
+          };
+          script = ''
+            set -eu
+
+            if ! [[ -e ${cfg.databasePasswordFile} ]]; then
+                echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                exit 1
+            fi
+
+            PSQL=${config.services.postgresql.package}/bin/psql
+
+            db_password="$(<'${cfg.databasePasswordFile}')"
+
+            $PSQL -tAc "SELECT 1 FROM pg_roles WHERE rolname='${cfg.databaseUsername}'" | grep -q 1 || $PSQL -tAc "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN PASSWORD '$db_password' CREATEDB"
+
+            $PSQL -tAc "SELECT 1 FROM pg_database WHERE datname = 'keycloak'" | grep -q 1 || $PSQL -tAc 'CREATE DATABASE "keycloak" OWNER "${cfg.databaseUsername}"'
+            current_owner=$($PSQL -tAc "SELECT pg_catalog.pg_get_userbyid(datdba) FROM pg_catalog.pg_database WHERE datname = 'keycloak'")
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
+                    exit 1
+                fi
+                touch "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+                $PSQL "keycloak" -tAc "REASSIGN OWNED BY \"$current_owner\" TO \"${cfg.databaseUsername}\""
+                rm "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+            fi
+          '';
+        };
+
+        systemd.services.keycloak = {
+          after = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          bindsTo = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          wantedBy = [ "multi-user.target" ];
+          environment = {
+            JBOSS_LOG_DIR = "/var/log/keycloak";
+            JBOSS_BASE_DIR = "/var/lib/keycloak";
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
```

Comment:
What are the bits that require this to be run as `root`? I feel like we might be able to get around this... :thinking:

Reply:
Yeah, I'm working on this right now. I _think_ the only part that needs root access is the ownership reassignment and permission changes to files, but I'll narrow it down to as few commands as possible and split the script into a privileged and an unprivileged part.

Reply:
It now uses root access only to get the secrets, which is a trade-off that afaik has to be made when `DynamicUser` is set.

Reply:
Correct.  However this will be fixed in future Systemd-versions  https://github.com/systemd/systemd/blob/master/NEWS#L320